src/Agent/NewRelic/Agent/Core/Aggregators/LogEventAggregator.cs

```diff
@@ -0,0 +1,207 @@
+
+        void CollectWithPriority(IList<LogEventWireModel> logEventWireModels, float priority);
+    }
+
+    /// <summary>
+    /// An service for collecting and managing logging events.
+    /// </summary>
+    public class LogEventAggregator : AbstractAggregator<LogEventWireModel>, ILogEventAggregator
+    {
+        private const String EntityType = "SERVICE";
+        private const String PluginType = "nr-dotnet-agent";
+        private const double ReservoirReductionSizeMultiplier = 0.5;
+
+        private readonly IAgentHealthReporter _agentHealthReporter;
+        private readonly ReaderWriterLockSlim _readerWriterLock = new ReaderWriterLockSlim();
+
+        private ConcurrentPriorityQueue<PrioritizedNode<LogEventWireModel>> _logEvents = new ConcurrentPriorityQueue<PrioritizedNode<LogEventWireModel>>(0);
+
+        public LogEventAggregator(IDataTransportService dataTransportService, IScheduler scheduler, IProcessStatic processStatic, IAgentHealthReporter agentHealthReporter)
+            : base(dataTransportService, scheduler, processStatic)
+        {
+            _agentHealthReporter = agentHealthReporter;
+            ResetCollections(_configuration.LogEventsMaximumPerPeriod);
+        }
+
+        protected override TimeSpan HarvestCycle => _configuration.LogEventsHarvestCycle;
+        protected override bool IsEnabled => _configuration.LogEventCollectorEnabled;
+
+        public override void Dispose()
+        {
+            base.Dispose();
+        }
+
+        public override void Collect(LogEventWireModel loggingEventWireModel)
+        {
+            _agentHealthReporter.ReportLoggingEventCollected();
+
+            _readerWriterLock.EnterReadLock();
+            try
+            {
+                AddEventToCollection(loggingEventWireModel);
+            }
+            finally
+            {
+                _readerWriterLock.ExitReadLock();
+            }
+        }
+
+        public void CollectWithPriority(IList<LogEventWireModel> logEventWireModels, float priority)
+        {
+            for (int i = 0; i < logEventWireModels.Count; i++)
+            {
+                _agentHealthReporter.ReportLoggingEventCollected();
+                logEventWireModels[i].Priority = priority;
+
+                _readerWriterLock.EnterReadLock();
+                try
+                {
+                    AddEventToCollection(logEventWireModels[i]);
+                }
+                finally
+                {
+                    _readerWriterLock.ExitReadLock();
+                }
+            }
+        }
+
+        protected override void Harvest()
+        {
+            ConcurrentPriorityQueue<PrioritizedNode<LogEventWireModel>> originalLogEvents;
+
+            _readerWriterLock.EnterWriteLock();
+            try
+            {
+                originalLogEvents = GetAndResetLogEvents(GetReservoirSize());
+            }
+            finally
+            {
+                _readerWriterLock.ExitWriteLock();
+            }
```

Comment:
My experience with InterlockedExchange is that you do not need this extra layer of locking. Remove the ReaderWriterLock, and this should work as expected.

Reply:
Pushed up a commit with locking removed.